File: greedy/cppRealizations/handofStraights.cpp
#include <vector>
#include <iostream>
#include <numeric>
#include <map>
#include <queue>
using namespace std;




class Solution {
public:
    bool isNStraightHand(vector<int>& hand, int groupSize) {
        if(hand.size() % groupSize != 0){
            return false;
        }
        map<int, int> count;
        for(int n : hand){
            count[n]++;
        }
        priority_queue<std::pair<int, int>, std::vector<std::pair<int, int>>, std::greater<std::pair<int, int>>> minH(count.begin(), count.end());
        while(minH.size() > 0){
            int first = minH.top().first;
            for(int i = first; i < first + groupSize; i++){
                if(count.find(i) == count.end()){
                    return false;
                }
                count[i]--;
                if(count[i] == 0){
                    if(i != minH.top().first){
                        return false;
                    }
                    minH.pop();
                }
            }

        }
        return true;
    }
};


ostream& operator<<(ostream& out, const vector<int>& str){
    out << "[";
    for(size_t i = 0; i < str.size(); i++){
        if(i < str.size() - 1){
            out << str[i] << ",";
        }else{
            out << str[i];
        }
    }
    out << "]";
    return out;
}

int main(){
    vector<int> hand = {1,2,3,6,2,3,4,7,8};
    cout << "Input: " << hand << endl;
    cout << "Solution: " << Solution().isNStraightHand(hand, 3) << endl;
    return 0;
}

File: greedy/handofStraights.py
import heapq

class Solution:
    def isNStraightHand(self, hand: list[int], groupSize: int) -> bool:
        if len(hand) % groupSize:
            return False
        
        count = {}
        for n in hand:
            count[n] = 1 + count.get(n, 0)

        minH = list(count.keys())
        heapq.heapify(minH)
        while minH:
            first = minH[0]
            for i in range(first, first + groupSize):
                if i not in count:
                    return False
                count[i] -= 1
                if count[i] == 0:
                    if i != minH[0]:
                        return False
                    heapq.heappop(minH)
        return True



if __name__ == '__main__':
    nums = [1,2,3,6,2,3,4,7,8]
    k = 3
    print("Input: " + str(nums) + ", " + str(k))
    print("Solution: " + str(Solution().isNStraightHand(nums, k)))
        